stan/math/prim/mat/fun/gp_matern52_cov.hpp

```diff
@@ -240,11 +247,13 @@ gp_matern52_cov(const std::vector<T_x1> &x1, const std::vector<T_x2> &x2,
  * @throw std::invalid_argument if length scale size != dimension of x1 or x2
  *
  */
-template <typename T_x1, typename T_x2, typename T_s, typename T_l>
+template <typename T_x1, typename T_x2, typename T_s, typename T_l, int R1,
+          int C1, int R2, int C2>
 inline typename Eigen::Matrix<
     typename stan::return_type<T_x1, T_x2, T_s, T_l>::type, Eigen::Dynamic,
     Eigen::Dynamic>
-gp_matern52_cov(const std::vector<T_x1> &x1, const std::vector<T_x2> &x2,
+gp_matern52_cov(const std::vector<Eigen::Matrix<T_x1, R1, C1>> &x1,
```

Comment:
Should R1 and R2 match? Should C1 and C2 match? Is there a case where they won't that shows up in Stan?

Reply:
yes, in fact, most of the time it won't math, and this is why I specified this here. Consider training X, and cross correlation between training and test X, where # of observations for the training and test sets are not the same. This is K(X, X_*) in  GPML. it's not a square matrix.

Reply:
actually this was incorrect, but this is just so that we can let `x[N]` be a vector of row vectors or column vectors.

Reply:
It's mostly so I can get the `divide_columns` function to work and has nothign to do with the training and test matrices above (i.e. to make the templating work in cmdstan I had to make the dimensions of the "matrix" flexible).